src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -15,37 +17,114 @@ class TimeoutCancellationTokenSourceWrapper : IDisposable
     {
         public TimeSpan Timeout { get; set; }
         CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+        CancellationToken _linkedToken;
 
-        public TimeoutCancellationTokenSourceWrapper() => Timeout = System.Threading.Timeout.InfiniteTimeSpan;
+#if DEBUG
+        bool _isRunning;
+#endif
+
+        public TimeoutCancellationTokenSourceWrapper() => Timeout = InfiniteTimeSpan;
 
         public TimeoutCancellationTokenSourceWrapper(TimeSpan timeout) => Timeout = timeout;
 
         /// <summary>
         /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
         /// and make sure that it hasn't been cancelled yet
         /// </summary>
-        public CancellationToken Start()
+        /// <param name="cancellationToken">An optional cancellation token that will be linked with the
+        /// contained <see cref="CancellationTokenSource"/></param>
+        /// <returns>The <see cref="CancellationToken"/> from the wrapped <see cref="CancellationTokenSource"/></returns>
+        public CancellationToken Start(CancellationToken cancellationToken = default)
         {
-            _timeoutCts.CancelAfter(Timeout);
-            if (_timeoutCts.IsCancellationRequested)
+#if DEBUG
+            Debug.Assert(!_isRunning);
+#endif
+
+            // If we already wrap a linked token source or we want to wrap a linked token source and the token to
+            // link isn't the same one that's already linked we always have to create a new CancellationTokenSource
+            if ((_linkedToken.CanBeCanceled || cancellationToken.CanBeCanceled) && _linkedToken != cancellationToken)
```

Comment:
Do we actually need `_linkedToken`? I think it's OK to assume that if the provided cancellation token can be cancelled, it's a new cancellation token that needs to be linked, and so we create a new CTS.

I also just realized another thing... Creating a new TCS (via CreateLinkedTokenSource) is a heap allocation. We could just as well register a callback instead ([CancellationToken.Register](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.register?view=netcore-3.1#System_Threading_CancellationToken_Register_System_Action_)), which only allocates a struct instead - that makes even more sense now that we have all the logic encapsulated in this type. Thoughts?

/cc @YohDeadfall @vonzshik

Reply:
> Do we actually need _linkedToken? I think it's OK to assume that if the provided cancellation token can be cancelled, it's a new cancellation token that needs to be linked, and so we create a new CTS.

Probably, yes.
Theoretically you could want to recycle a linked token source if it is already linked to the same token but I guess I made up that use case and it doesn't happen IRL or at least not often enough to worry about the extra allocation.

> I also just realized another thing... Creating a new TCS (via CreateLinkedTokenSource) is a heap allocation. We could just as well register a callback instead (CancellationToken.Register), which only allocates a struct instead - that makes even more sense now that we have all the logic encapsulated in this type. Thoughts?

Yeah, I like the Idea.
If I understand it correctly the linked source is not doing very much besides that itself and in that case we could save an allocation that's happening every time someone passes a cancellable token into a method.

Reply:
While thinking about this the question comes up, when we'd want to dispose or unregister the registration.
The point where you call `Stop()` is most likely also the point where you want to unregister the callback from the passed token, but is this always the case?
If not we'd need to expose the registration outside of the wrapper, possibly returning it from `Start()` and `Reset()` instead of the token.